diff -crB ubercart-orig/uc_cart/uc_cart.module ubercart/uc_cart/uc_cart.module
*** ubercart-orig/uc_cart/uc_cart.module	2013-04-26 19:51:06.013182172 -0600
--- ubercart/uc_cart/uc_cart.module	2013-04-26 19:51:17.356181657 -0600
***************
*** 1319,1328 ****
  function uc_cart_get_id($create = TRUE) {
    global $user;
  
    if ($user->uid) {
      return $user->uid;
    }
!   elseif (!isset($_SESSION['uc_cart_id']) && $create) {
      $_SESSION['uc_cart_id'] = md5(uniqid(rand(), TRUE));
    }
  
--- 1319,1331 ----
  function uc_cart_get_id($create = TRUE) {
    global $user;
  
+ /*
    if ($user->uid) {
      return $user->uid;
    }
!   elseif
! */
!   if (!isset($_SESSION['uc_cart_id']) && $create) {
      $_SESSION['uc_cart_id'] = md5(uniqid(rand(), TRUE));
    }
  
Binary files ubercart-orig/uc_cart/.uc_cart.module.swp and ubercart/uc_cart/.uc_cart.module.swp differ
